core/network/impl/protocols/light.cpp

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "network/impl/protocols/light.hpp"
+
+#include "network/common.hpp"
+#include "runtime/common/trie_storage_provider_impl.hpp"
+#include "runtime/runtime_environment_factory.hpp"
+
+namespace kagome::network {
+  LightProtocol::LightProtocol(
+      libp2p::Host &host,
+      const application::ChainSpec &chain_spec,
+      const primitives::GenesisBlockHeader &genesis,
+      std::shared_ptr<blockchain::BlockHeaderRepository> repository,
+      std::shared_ptr<storage::trie::TrieStorage> storage,
+      std::shared_ptr<runtime::RuntimeEnvironmentFactory> env_factory)
+      : RequestResponseProtocolType{
+          kName,
+          host,
+          make_protocols(kLightProtocol, chain_spec, genesis),
+          log::createLogger(kName),
+      },
+      repository_{std::move(repository)},
+      storage_{std::move(storage)},
+      env_factory_{std::move(env_factory)} {}
+
+  outcome::result<LightProtocol::ResponseType> LightProtocol::onRxRequest(
+      RequestType req, std::shared_ptr<Stream>) {
+    std::unordered_set<common::Buffer> proof;
+    auto prove = [&](common::BufferView raw) { proof.emplace(raw); };
+    OUTCOME_TRY(header, repository_->getBlockHeader(req.block));
+    OUTCOME_TRY(batch,
+                storage_->getProofReaderBatchAt(header.state_root, prove));
+    auto call = boost::get<LightProtocolRequest::Call>(&req.op);
+    if (call) {
+      OUTCOME_TRY(factory, env_factory_->start(req.block));
+      OUTCOME_TRY(env, factory->withStorageBatch(std::move(batch)).make());
+      OUTCOME_TRY(
+          env->module_instance->callExportFunction(call->method, call->args));
```

Comment:
You could probably add a method to raw executor which would accept a runtime environment to keep this logic inside the executor class.

Reply:
I don't see need for new function which is used only here yet